### DISCUSSION TITLE
Chatwoot Integration Guide

### DISCUSSION THREAD

**Discussion by smashah** [open-wa/wa-automate-nodejs#2371] [General]
# Setup

TODO
- [ ]  Fix the URL in the chatwoot inbox setup video. It should be `.../chatwoot`
- [ ] Show how to add the api key to chatwoot inbox setup. `.../chatwoot?api_key=xxxxxxxx`

 
## Step 1

Create a new inbox with the API channel. 

![CleanShot 2022-03-08 at 22 16 16](https://user-images.githubusercontent.com/5903821/157335355-52342b13-aa87-430b-88fd-8c363ef1b653.gif)

> 	If you already know your EASY API host, then set the webhook now (e.g if your easy api is on `http://localhost:8006` then set the webhook `http://localhost:8006/chatwoot`. You can do this later also.

## Step 2

Go to your newly created inbox in the `conversations` tab. Note down the url, use that to convert it to the chatwoot api url.

> `http://localhost:3000/app/accounts/1/inbox/1` becomes `http://localhost:3000/api/v1/accounts/1/inboxes/1` - basically replace `app` with `api/v1`

https://user-images.githubusercontent.com/5903821/157480906-ef0f14a4-0c59-4da7-a433-6f166e1d582b.mp4

## Step 3

Grab the `access token` from the bottom of the `Profile settings` page. E.g `qZgaibhBPQQqddtGUeytsFqu`

## Step 4

Start the easy API with `--chatwoot-url="http://localhost:3000/api/v1/accounts/1/inboxes/1"` and `--chatwoot-api-access-token=qZgaibhBPQQqddtGUeytsFqu`

## Step 5

If you haven't already, set the webhook url for your EASY API instance on the new inbox `Settings > Inboxes > [Your inbox] > Settings > WEBHOOK URL`

extra media:

![CleanShot 2022-03-08 at 22 11 10@2x](https://user-images.githubusercontent.com/5903821/157334471-fc073405-2788-4a2b-8879-d904882f8ae3.png)

> [!TIP]
> If you are running your chatwoot docker container in the same host as your sessions then you need to make sure the chatwoot instance can "reach" your sessions by exposing localhost to your docker container and then informing your easy api of the api host to expect requests on.

add this line to the rails and worker services in the chatwoot docker-compose yaml:

```
    extra_hosts:
      - "host.docker.internal:host-gateway"
```

Now lets say for example your session is running on port 8006 in the host machine, your inbox webhook url will be `http://host.docker.internal:8006/chatwoot`

If you have further issues then you need to inform easy api of the api host that it is now being addressed by, you can do this by adding the cli flag `--api-host http://host.docker.internal` or `apiHost: "http://host.docker.internal"`

**Comment by trcalgaro** [comment]
I managed to make the connection and I see the messages generated in wa-automate, but when I answer in chatwood the message does not return to wa-automate, is there this possibility?

**Reply by smashah** [reply]
@trcalgaro are you sure you followed the instructions fully? You have to set the webhook in the chatwoot inbox to the address of the open-wa EASY API instance. If you set it to localhost and chatwoot is running inside docker then it may not be able to send requests to open-wa due to networking issues. In which case you have to give chatwoot access to the host network or you need to expose your open-wa instance using [`ngrok`](https://ngrok.com/), [`cloudflared`](https://www.cloudflare.com/products/tunnel/) or the `--tunnel` flag

**Reply by trcalgaro** [reply]
Hi @smashah 
Yes, I followed all the configuration phases, including the placement of the webhook, but I have an apiKey  `--key`   configured in the EASY API instance, as there is no way to configure the apiKey in the WebHook configuration, I believe that is why it should not be returning.

My instances are running on a local VMware, so they are all in the same IP range and can communicate with each other.

Is there any way I can make the webhook open-wa EASY API without apiKey so both can communicate? at least for testing?

**Reply by smashah** [reply]
@trcalgaro absolutely there is (it's the default behaviour to start without api key).

Can you show me your exact EASY API command (and/or the `cli.config.json` file)

**Reply by trcalgaro** [reply]
Hello, I didn't find the file "cli.config.json" below follows command to start the tool.

I keep getting messages but when I reply they don't return to wa-automate

![image](https://user-images.githubusercontent.com/41573074/163293119-f54e7dfc-b96e-43c3-9a06-19ed25bb41e7.png)

`npx @open-wa/wa-automate --multi-device --port 8101 --chatwoot-url "http://170.10.10.132:3000/api/v1/accounts/1/inboxes/1" --chatwoot-api-access-token=XXxxxxxxxxxxxxxxxxxxx --license-key xxxxxxxxxxxxxxxxxxxxx --bot-press-url http://xxxxxxxxxxxxx/api/v1/bots/bom-gosto/converse  --key xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx --config "/home/xxxxxxx/WhatsApp/BG/config.json" --webhook https://xxxxxxxx/api/webhook/xxxxxxxxxxxxxxxxxx --popup --socket --no-kill-on-logout --session-id BG --no-kill-on-logout true --ev https://xxxxxxxxxxxxx/api/webhook/xxxxxxxxxxxxxxxxx --use-chrome --discord string xxxxxxxxxxxxxxxxxxxxxxxxx`

![image](https://user-images.githubusercontent.com/41573074/163292284-33fd4d89-0f58-41b9-81fa-7188458bc521.png)

**Reply by smashah** [reply]
You can create a cli.config.json file to simplify your CLI command.

Your CLI command can be converted to:

`cli.config.json`
```js
{
  "multiDevice": true,
  "port": 8101,
  "chatwootUrl": "http://170.10.10.132:3000/api/v1/accounts/1/inboxes/1",
  "chatwootApiAccessToken": "XXxxxxxxxxxxxxxxxxxxx",
  "licenseKey": "xxxxxxxxxxxxxxxxxxxxx",
  "botPressUrl": "http://xxxxxxxxxxxxx/api/v1/bots/bom-gosto/converse",
  "webhook": "https://xxxxxxxx/api/webhook/xxxxxxxxxxxxxxxxxx",
  "popup": true,
  "socket": true,
  "noKillOnLogout": true,
  "sessionId": "BG",
  "ev": "https://xxxxxxxx/api/webhook/xxxxxxxxxxxxxxxxxx",
  "useChrome": true,
  "discord": "xxxxxxxxxxxxxxxxxxxxxxxxx"
}
```

then you can run the cli without any flags because`cli.config.json` is automatically picked up, or you can set the filepath with `--config "/home/xxxxxxx/WhatsApp/BG/config.json"` where you can add the above JSON.

```bash
# picks up cli.config.json in the current directory automatically
> npx @open-wa/wa-automate 

#or use /home/xxxxxxx/WhatsApp/BG/config.json
> npx @open-wa/wa-automate --config "/home/xxxxxxx/WhatsApp/BG/config.json"
```

As you have noticed, I removed "key" from the json. Chatwoot can't handle the api key right now. I might make an issue where the api key is taken from a query param specifically for the `/chatwoot` endpoint - but for now it's best to keep it empty when integrating with chatwoot.

**Reply by trcalgaro** [reply]
@smashah 
Thanks for the tip

I used your recommendations in the cli.config file but still the messages still don't return from chatwoot to wa

I tried restarting the chatwoot service anyway, without success!

**Reply by smashah** [reply]
@trcalgaro I just noticed, in your screenshot you set the url in chatwoot as `http://170.10.10.127:8101`

That is the address of the instance, you need to add the specific endpoint `/chatwoot`.

Change it to:

`http://170.10.10.127:8101/chatwoot`

**Reply by trcalgaro** [reply]
Now yes... returned

Thanks for the help, it would be important to put the proper path in the inbox webhook in the tutorial.

I tested here adding the "key" again and it doesn't really work with it. I think you can handle the error yes, since it would be important to keep the key for other communications via socket (Node Red for example), or swagger API

**Reply by smashah** [reply]
@trcalgaro 

Seems as though I have forgotten that I already did this:

https://github.com/open-wa/wa-automate-nodejs/blob/33568e00728085a6cb0d8a61c7547e05f74778e6/src/cli/server.ts#L53


So this should work in chatwoot inbox settings: 

`http://170.10.10.127:8101/chatwoot?api_key=xxxxxxxxxxxxxxxx`

**Reply by trcalgaro** [reply]
Tks, its Working!

**Comment by hasanphd** [comment]
Hi @smashah I tried all that's said above and tried to configure Chatwoot with WhatsApp. 
However, failed to make the outgoing messages work. Incoming of messages is working fine. 

I couldn't figure out the config.js correctly so deleted that. 
![Screenshot 2022-06-05 161252](https://user-images.githubusercontent.com/8126083/172047913-827493d9-8813-4b9c-946b-48ccdc2d3be4.png)

Command I am using:

` npx @open-wa/wa-automate -p 8082 --chatwoot-url "http://127.0.0.1:3000/app/accounts/1/inbox/2" --chatwoot-api-access-token xxxxxxxxxxxxxxxxxx`

**Reply by smashah** [reply]
@hasanphd Does your chatwoot container have access to the host network? Try using ngrok so that chatwoot can access open-wa 

`ngrok http 8082` then use the url from that inside chatwoot

**Reply by hasanphd** [reply]
Actually, both chatwoot and open-wa are on Localhost. Chatwoot running in WSL. And open-wa via Windows. 
Both localhost URLs work fine, but outgoing messages does not work that are sent from Chatwoot API call to open-wa. 

I checked and it does not seem to be port issue. I will however deploy chatwoot to ngrok too and see meanwhile you get a chance to respond here.

**Reply by smashah** [reply]
Just because you can access both on localhost doesn't mean chatwoot has access to the open-wa port. If chatwoot is running within docker and based on your setup, you can access chatwoot on localhost but your chatwoot instance cannot connect to the machine's localhost network to access port 8082 to send requests to open-wa. My advice was to put an ngrok infront of open-wa then update your inbox settings with the ngrok url - not the other way around.

If you want to test your open-wa API you can use the `/api-docs/`

**Reply by hasanphd** [reply]
It worked! Genius!

**Reply by fejorca** [reply]
I'm having this same issue, wa-automate is running on a docker container and messages are going to the chatwoot inbox but can't respond to any message, wa-automate docker container has a public address but when trying to use the url with "/chatwoot" it shows a message that says "Cannot GET /chatwoot", also I can see api-docs and test without issues but can't get chatwoot message responses

EDIT: I'm using NPM with let's encrypt certificates

**Reply by smashah** [reply]
@fejorca `/chatwoot` is solely a `POST` path designed for chatwoot - not for a front end user.

The problem is that chatwoot docker compose setup is running on it's own network - it doesn't have access to the host network.

Please follow this issue for updates:

https://github.com/open-wa/wa-automate-nodejs/issues/2869

**Comment by smashah** [comment]
@fejorca @hasanphd @trcalgaro @apnerve 


Please track this issue for massive chatwoot integration QoL improvements:

https://github.com/open-wa/wa-automate-nodejs/issues/2869

**Comment by lazmo88** [comment]
@smashah just got this up and running and so far so good. Only issue I noticed is that Chatwoot won't receive replies sent from other devices or WA web UI. Is this normal or expected behaviour? It would be great if all conversation history could stay in sync also on CW, including replies agents sent from WA web or WA mobile apps.
